diff --git a/WowPacketParserModule.V5_3_0_16981/Parsers/MiscellaneousHandler.cs b/WowPacketParserModule.V5_3_0_16981/Parsers/MiscellaneousHandler.cs
index e422a6b..c460b12 100644
--- a/WowPacketParserModule.V5_3_0_16981/Parsers/MiscellaneousHandler.cs
+++ b/WowPacketParserModule.V5_3_0_16981/Parsers/MiscellaneousHandler.cs
@@ -65,26 +65,17 @@ namespace WowPacketParserModule.V5_3_0_16981.Parsers
         {
             var guid = new byte[8];
 
-            guid[2] = packet.ReadBit();
-            guid[3] = packet.ReadBit();
-            guid[6] = packet.ReadBit();
-            guid[0] = packet.ReadBit();
-            guid[1] = packet.ReadBit();
-            guid[5] = packet.ReadBit();
-            guid[4] = packet.ReadBit();
-            guid[7] = packet.ReadBit();
-
-
-            packet.ReadXORByte(guid, 1);
-            packet.ReadXORByte(guid, 2);
-            packet.ReadXORByte(guid, 6);
-            packet.ReadXORByte(guid, 4);
-            packet.ReadXORByte(guid, 7);
-            packet.ReadXORByte(guid, 0);
-            packet.ReadXORByte(guid, 3);
-            packet.ReadXORByte(guid, 5);
-
+            packet.StartBitStream(guid, 2, 3, 6, 0, 1, 5, 4, 7);
+            packet.ParseBitStream(guid, 1, 2, 6, 4, 7, 0, 3, 5);
             packet.WriteGuid("Player GUID: ", guid);
         }
+
+        [Parser(Opcode.SMSG_WEATHER)]
+        public static void HandleWeatherStatus(Packet packet)
+        {
+            packet.ReadEnum<WeatherState>("State", TypeCode.Int32);
+            packet.ReadSingle("Grade");
+            packet.ReadBit("Unk bit");
+        }
     }
 }
diff --git a/WowPacketParserModule.V5_3_0_16981/Parsers/NpcHandler.cs b/WowPacketParserModule.V5_3_0_16981/Parsers/NpcHandler.cs
index ef8a2bf..fb5fe79 100644
--- a/WowPacketParserModule.V5_3_0_16981/Parsers/NpcHandler.cs
+++ b/WowPacketParserModule.V5_3_0_16981/Parsers/NpcHandler.cs
@@ -1,9 +1,13 @@
 using System;
+using System.Collections.Generic;
+using System.Globalization;
 using System.Text;
 using WowPacketParser.Enums;
 using WowPacketParser.Misc;
 using WowPacketParser.Parsing;
 using WowPacketParser.Store;
+using WowPacketParser.Store.Objects;
+using Guid = WowPacketParser.Misc.Guid;
 
 namespace WowPacketParserModule.V5_3_0_16981.Parsers
 {
@@ -40,30 +44,94 @@ namespace WowPacketParserModule.V5_3_0_16981.Parsers
             var guid = new byte[8];
             var menuEntry = packet.ReadUInt32("Menu Id");
             var gossipId = packet.ReadUInt32("Gossip Id");
-            guid[2] = packet.ReadBit();
-            guid[0] = packet.ReadBit();
-            guid[6] = packet.ReadBit();
-            guid[4] = packet.ReadBit();
-            guid[1] = packet.ReadBit();
-            guid[5] = packet.ReadBit();
-            guid[3] = packet.ReadBit();
-            guid[7] = packet.ReadBit();
+            packet.StartBitStream(guid, 2, 0, 6, 4, 1, 5, 3, 7);
 
             var bits8 = packet.ReadBits(8);
 
             packet.ReadXORByte(guid, 0);
-
             packet.ReadWoWString("Box Text", bits8);
-            packet.ReadXORByte(guid, 6);
-            packet.ReadXORByte(guid, 5);
-            packet.ReadXORByte(guid, 7);
-            packet.ReadXORByte(guid, 4);
-            packet.ReadXORByte(guid, 3);
-            packet.ReadXORByte(guid, 2);
-            packet.ReadXORByte(guid, 1);
+            packet.ReadXORBytes(guid, 6, 5, 7, 4, 3, 2, 1);
 
             Storage.GossipSelects.Add(Tuple.Create(menuEntry, gossipId), null, packet.TimeSpan);
             packet.WriteGuid("GUID", guid);
         }
+
+        [HasSniffData]
+        [Parser(Opcode.SMSG_GOSSIP_MESSAGE)]
+        public static void HandleNpcGossip(Packet packet)
+        {
+            var guid = new byte[8];
+            uint[] titleLen;
+            uint[] OptionTextLen;
+            uint[] BoxTextLen;
+
+            packet.StartBitStream(guid, 0, 1);
+            var questgossips = packet.ReadBits("Amount of Quest gossips", 19);
+            guid[2] = packet.ReadBit();
+            var AmountOfOptions = packet.ReadBits("Amount of Options", 20);
+
+            titleLen = new uint[questgossips];
+            for (var i = 0; i < questgossips; ++i)
+            {
+                packet.ReadBit("Change Icon", i);
+                titleLen[i] = packet.ReadBits(9);
+            }
+            guid[3] = packet.ReadBit();
+
+            OptionTextLen = new uint[AmountOfOptions];
+            BoxTextLen = new uint[AmountOfOptions];
+            for (var i = 0; i < AmountOfOptions; ++i)
+            {
+                BoxTextLen[i] = packet.ReadBits(12);
+                OptionTextLen[i] = packet.ReadBits(12);
+            }
+
+            packet.StartBitStream(guid, 5, 4, 6, 7);
+            packet.ResetBitReader();
+            packet.ReadXORByte(guid, 6);
+
+            var gossip = new Gossip();
+
+            gossip.GossipOptions = new List<GossipOption>((int)AmountOfOptions);
+            for (var i = 0; i < AmountOfOptions; ++i)
+            {
+                var gossipOption = new GossipOption
+                {
+                    OptionText = packet.ReadWoWString("Text", OptionTextLen[i], i),
+                    BoxText = packet.ReadWoWString("Box Text", BoxTextLen[i], i),
+                    RequiredMoney = packet.ReadUInt32("Required money", i),
+                    OptionIcon = packet.ReadEnum<GossipOptionIcon>("Icon", TypeCode.Byte, i),
+                    Box = packet.ReadBoolean("Box", i),
+                    Index = packet.ReadUInt32("Index", i)
+                };
+
+                gossip.GossipOptions.Add(gossipOption);
+            }
+            packet.ReadXORByte(guid, 0);
+
+            for (var i = 0; i < questgossips; i++)
+            {
+                packet.ReadEnum<QuestFlags>("Flags", TypeCode.UInt32, i);
+                packet.ReadWoWString("Title", titleLen[i], i);
+                packet.ReadUInt32("Icon", i);
+                packet.ReadEnum<QuestFlags2>("Flags 2", TypeCode.UInt32, i);
+                packet.ReadInt32("Level", i);
+                packet.ReadEntryWithName<UInt32>(StoreNameType.Quest, "Quest ID", i);
+            }
+
+            var textId = packet.ReadUInt32("Text Id");
+            packet.ReadXORBytes(guid, 4, 3);
+            var menuId = packet.ReadUInt32("Menu Id");
+            packet.ReadUInt32("Friendship Faction");
+            packet.ReadXORBytes(guid, 7, 1, 5, 2);
+            packet.WriteGuid("GUID", guid);
+
+            var GUID = new Guid(BitConverter.ToUInt64(guid, 0));
+            gossip.ObjectType = GUID.GetObjectType();
+            gossip.ObjectEntry = GUID.GetEntry();
+
+            Storage.Gossips.Add(Tuple.Create(menuId, textId), gossip, packet.TimeSpan);
+            packet.AddSniffData(StoreNameType.Gossip, (int)menuId, GUID.GetEntry().ToString(CultureInfo.InvariantCulture));
+        }
     }
 }
diff --git a/WowPacketParserModule.V5_3_0_16981/Parsers/SpellHandler.cs b/WowPacketParserModule.V5_3_0_16981/Parsers/SpellHandler.cs
index be4d1a3..ce4dc1f 100644
--- a/WowPacketParserModule.V5_3_0_16981/Parsers/SpellHandler.cs
+++ b/WowPacketParserModule.V5_3_0_16981/Parsers/SpellHandler.cs
@@ -685,6 +685,5 @@ namespace WowPacketParserModule.V5_3_0_16981.Parsers
             packet.WriteGuid("GUID10", guid10);
             packet.WriteGuid("GUID11", guid11);
         }
-
     }
 }
